tlog-witness.md

```diff
@@ -0,0 +1,226 @@
+    20852163
+    CsUYapGGPo4dkMgIAUqom/Xajj7h2fB2MPA3j2jxq2I=
+
+    — example.com/behind-the-sofa Az3grlgtzPICa5OS8npVmf1Myq/5IZniMp+ZJurmRDeOoRDe4URYN7u5/Zhcyv2q1gGzGku9nTo+zyWE+xeMcTOAYQ8=
+    — example.com/behind-the-sofa opLqBQsREYCgu6xQkYQwJr9fo45a62DN9EdmHXnZdXNqlcVGlCum2Wks+49/V6267UEjw6QUXTS5Rovnzv++qbSzm9Q=
+
+The witness MUST verify the checkpoint signature against the public key(s) it
+trusts for the checkpoint origin, and it MUST ignore signature from unknown
+keys. If the checkpoint origin is unknown, the witness MUST respond with a "404
+Not Found" HTTP status code. If none of the signatures verify against a trusted
+public key, the witness MUST respond with a "403 Forbidden" HTTP status code.
+
+The previous size MUST be equal to or lower than the checkpoint size. Otherwise,
+the witness MUST respond with a "400 Bad Request" HTTP status code.
+
+The witness MUST check that the previous size matches the size of the latest
+checkpoint it cosigned for the checkpoint's origin (or zero if it never cosigned
+a checkpoint for that origin). If it doesn't match, the witness MUST respond
+with a "409 Conflict" HTTP status code. The response body MUST consist of the
+tree size of the latest cosigned checkpoint in decimal, followed by a newline
+(U+000A). The response MUST have a `Content-Type` of `text/x.tlog.size`.
+
+If a client doesn't have information on the latest cosigned checkpoint, it MAY
+initially make a request with a previous size of zero to obtain it.
+
+The consistency proof lines MUST encode a Merkle Consistency Proof from the
+previous size to the checkpoint size according to [RFC 6962][], Section 2.1.2. The
+proof MUST be empty only if the previous size is zero or if it matches the size
+in the checkpoint. If the Merkle Consistency Proof doesn't verify, the witness
+MUST respond with a "422 Unprocessable Entity" HTTP status code.
+
+If the origin is known, and the signature is valid, the witness SHOULD log the
+request even if the consistency proof doesn't verify (but MUST NOT cosign the
+checkpoint) as it might be proof of log misbehavior.
+
+If all the checks above pass, the witness MUST update its record of the latest
+cosigned checkpoint and respond with a "200 Success" HTTP status code. The
+response body MUST be a sequence of one or more [note][] signature lines, each
+starting with the `—` character (U+2014) and ending with a newline character
+(U+000A). The signatures MUST be [cosignatures][] from the witness key(s) on the
+checkpoint.
+
+    — witness.example/w1 CMp+6LWBU0anHGH5aNDTJkH/gj79sG+T6+iP2ThYN5krrDJbR1HDnucjL39QsZTSvVjyQLrdk3DXDqI5G2HgLatVs0pWh6Up69HVOw==
+    — witness.example/w1 I7rEps0pvK2UqkS2gSpVUDhrhVtQV9lgF6pRrWAvjJHjyWpW7VcE3SiOlVlbQNt64vWhO+DlkL0+UfzuOBMh9ChdMkP1vi/lCAsmlw==
+    — witness.example/w2 AWui8Sk55XjYLOijihBjhqEH6nS1ndDymE0a+6idX7pLcnoB+dhnz0854aLZgrrKbYKA7nC3HNJhm/kWl7oJlqU3rXXvpysAdyP3wQ==
+
+The client MUST ignore any cosignatures from unknown keys. To parse the
```

Comment:
This may be out of scope for the spec, but how do we expect clients to discover witness keys?

Reply:
It will depend on who sets policy. By example, the Go sumdb client policy is set by the Go team, who sets the defaults of the go tool in the open source project. I expect in most ecosystems the log operators will also set client policy (in the open, and in such a way that the relevant community trusts the resulting transparency system).